files_orpg_sw/src/cpc100/lib004/str.3
str(3)			String manipulation routines			str(3)

NAME
STR - string manipulation routines

SYNOPSIS

char *STR_grow (char *str, int length);

char *STR_copy (char *dest_str, const char *src_str);

char *STR_cat (char *dest_str, const char *src_str);

char *STR_append (char *dest_str, const char *src_str, int size);

char *STR_replace (char *str, int offset, int len, char *c_str, int size);

int STR_size (char *str);

char *STR_reset (char *str, int length);

void STR_free (char *str);

char *STR_gen (char *str, ...);

DESCRIPTION

    The STR routines are convenient functions for processing variable-sized
    character strings. They provide support of string operations (such as
    copying, concatenating and sub-string replacement) with automatic memory
    reallocation and, thus, free the calling application code from managing the
    buffers. These routines use MISC_malloc so exception handling is not
    required by the calling functions. That is the STR functions never return
    with failure. If one does not want to terminate the process upon malloc
    failure, one can call MISC_malloc_retry to let MISC_malloc retry until
    success. When extending the buffer size, STR allocates extra space so
    memory reallocation is not needed for every string expansion. STR functions
    are event interrupt-safe.

    A pointer returned by any of the STR functions are called a STR pointer. If
    a STR string is a pointer to a null terminated string, any of the standard
    C string manipulation functions that do not modify a string length may be
    used with STR pointers (i.e. strstr, strchr, etc). A STR pointer, however,
    cannot be freed by "free". "STR_free", instead, must be used.
	
    STR_grow increases the allocated string buffer size of STR pointer "str"
    to at least "length" bytes. "str" must be NULL or a STR pointer returned by
    one of the STR functions. If "str" is NULL, a new empty STR pointer with a
    buffer of "length" bytes will be created. This function returns the
    requested STR pointer. STR_grow can be used for initializing the buffer
    to a desired size.

    STR_copy copies "src_str" to "dest_str" which must be a STR pointer
    returned from STR functions. "src_str" can be either a STR string or a C
    string. If "dest_str" is NULL, a new STR pointer is created and returned.
    "src_str" must be either a null-terminated string or NULL. If it is NULL,
    all data bytes are discarded and the number of data bytes is set to 0. The
    function returns the STR pointer that contains the result. STR_copy (str,
    NULL) can be used for deleting the data in "src_str".

    STR_cat appends "src_str" to "dest_str" which must be a STR pointer
    returned from STR functions. "dest_str" must contain null-terminated
    string. "src_str" can be either a STR string or a C string. If "dest_str"
    is NULL, a new STR pointer is created and returned. The function returns
    the STR pointer that contains the result. STR_cat should only be used for
    null-terminated strings.

    STR_append appends "size" bytes pointed by "src_str" to the end of
    "dest_str" which must be a STR pointer returned from STR functions.
    "src_str" can be either a STR string or a C string. If "dest_str" is NULL,
    a new STR pointer is created and returned. The function returns the STR
    pointer that contains the result. STR_append allows the STR module to be
    used for binary (not null-terminated) strings.

    STR_replace replaces, with "c_str" of "size" bytes, "len" bytes at offset
    "offset" of string pointed to by "str" which must be a STR pointer returned
    from STR functions. "c_str", however, can be either a STR string or a C
    string. If cases of NULL "str", invalid "len" and "offset", STR_replace
    returns "str" without any processing. The function returns the STR pointer
    that contains the result.

    STR_free frees STR pointer "str" which must be a STR pointer returned from
    STR functions.

    STR_size returns the number of bytes currently stored in "str" which must 
    be a STR pointer returned from STR functions. 

    STR_reset sets/resets the buffer size to at least "length" if "length" is
    non-zero and discards all data (sets number of data bytes to 0). If "str" 
    is NULL, a new STR pointer is created and returned.

    STR_create is for backward compatibility. It is equivalent to STR_copy
    (NULL, orig_str).

    STR_gen appends a list of strings to generate a string. The first argument,
    which must be a STR pointer, specifies the place to put the generated 
    string. Any content in "str" is discarded. The argument list must be 
    terminated by "NULL". STR_gen returns the first argument.


EXAMPLES:

	#include <infr.h>

        /*  Create the initial string  */
	char* str = STR_copy (NULL, "The fox jumped over");

	/*  Memory allocation will happen automatically if needed */
	str = STR_cat (str, "the moon");

	/*  Find the word moon in a STR created string */
	if (strstr (str, "moon") != NULL)
            printf ("The string contains the word moon\n");

	/*  Free the string  */
	STR_free (str);

	/*  Create a new string that contains 250 bytes of space */
	str = STR_grow (NULL, 250);

	/*  Concatenate some stuff */
	str = STR_copy (str, "Some stuff");

	/*  Print the string out */
	printf ("The string equals %s\n", str);

	/*  Increase the length str buffer to 275  */
	str = STR_grow (str, 275);

	STR_free (str);



files_orpg_sw/src/cpc100/lib004/str.c

/*******************************************************************

    Description: The variable size string processing module (STR).

*******************************************************************/

/*
 * RCS info
 * $Author: jing $
 * $Locker:  $
 * $Date: 2012/12/06 22:35:19 $
 * $Id: str.c,v 1.12 2012/12/06 22:35:19 jing Exp $
 * $Revision: 1.12 $
 * $State: Exp $
 */

#include <config.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#include <infr.h>
#include <str.h>

typedef struct {
    int b_size;			/* buffer size */
    int n_bytes;		/* number of data bytes stored */
} Str_hd_t;			/* header for STR pointer */


static char *Str_expand (void *str, int length);

/***********************************************************************

    Increases the buffer size of string "str" to "length".

***********************************************************************/

char *STR_grow (void *vstr, int length) {
    Str_hd_t *hd, *new_hd;
    char *p, *str;

    str = (char *)vstr;
    hd = NULL;			/* not necessary - to turn off gcc warning */
    if (str != NULL) {
	hd = (Str_hd_t *)(str - sizeof (Str_hd_t));
	if (hd->b_size >= length)
	    return (str);
    }

    p = (char *)MISC_malloc (length + sizeof (Str_hd_t));
    new_hd = (Str_hd_t *)p;
    p += sizeof (Str_hd_t);
    new_hd->b_size = length;
    if (str == NULL)
	new_hd->n_bytes = 0;
    else {
	memcpy (p, str, hd->n_bytes);
	new_hd->n_bytes = hd->n_bytes;
	MISC_free (str - sizeof (Str_hd_t));
    }
    return (p);
}

/***********************************************************************

    This is a duplication of STR_copy.

***********************************************************************/

char *STR_create (const void *orig_str) {

    return (STR_copy (NULL, (char *)orig_str));
}

/***********************************************************************

    Grows the buffer size and discards all data.

***********************************************************************/

char *STR_reset (void *vstr, int length) {
    char *str = (char *)vstr;
    if (str != NULL)
	((Str_hd_t *)(str - sizeof (Str_hd_t)))->n_bytes = 0;
    if (length > 0 || str == NULL)
	str = STR_grow (str, length);
    return (str);
}

/***********************************************************************

    Copies string from "src_str" to "dest_str". "dest_str" can be NULL.

***********************************************************************/

char *STR_copy (char *dest_str, const char *src_str) {
    int nb = 0;
    if (src_str != NULL)
	nb = strlen (src_str) + 1; 
    dest_str = STR_grow (dest_str, nb);
    if (src_str != NULL)
	strcpy (dest_str, src_str);
    ((Str_hd_t *)(dest_str - sizeof (Str_hd_t)))->n_bytes = nb;
    return (dest_str);
}

/***********************************************************************

    Concatenats string "src_str" to "dest_str". "dest_str" can be NULL.
    "src_str" cannot be NULL.

***********************************************************************/

char *STR_cat (char *dest_str, const char *src_str) {
    int nb;

    if (dest_str == NULL || 
	((Str_hd_t *)(dest_str - sizeof (Str_hd_t)))->n_bytes == 0)
	return (STR_copy (dest_str, src_str));
    nb = strlen (dest_str) + strlen (src_str) + 1;
    dest_str = Str_expand (dest_str, nb);
    strcat (dest_str, src_str);
    ((Str_hd_t *)(dest_str - sizeof (Str_hd_t)))->n_bytes = nb;
    return (dest_str);
}

/***********************************************************************

    Appends "size" bytes pointed by "src_str" to the end of "dest_str". 
    "dest_str" can be NULL.

***********************************************************************/

char *STR_append (void *dest_str, const void *src_str, int size) {
    int nb = 0;
    if (dest_str != NULL)
	nb = ((Str_hd_t *)((char *)dest_str - sizeof (Str_hd_t)))->n_bytes;
    dest_str = Str_expand (dest_str, nb + size);
    if (src_str != NULL)
	memcpy ((char *)dest_str + nb, (char *)src_str, size);
    ((Str_hd_t *)((char *)dest_str - sizeof (Str_hd_t)))->n_bytes = nb + size;
    return ((char *)dest_str);
}

/************************************************************************

    Replaces data at "offset" of "len" bytes in string "str" by C "c_str".

************************************************************************/

char *STR_replace (char *str, int offset, int len, char *c_str, int size) {
    int s_size, nb, s;

    if (str == NULL || offset < 0 || len < 0 ||
	offset > (s_size = ((Str_hd_t *)(str - sizeof (Str_hd_t)))->n_bytes))
	return (str);			/* not done */

    nb = s_size + size - len;
    str = Str_expand (str, nb);
    if (offset + len > s_size)
	len = s_size - offset;
    s = s_size - (offset + len);
    if (s > 0 && size != len)
	memmove (str + offset + size, str + offset + len, s);
    if (size > 0)
	memcpy (str + offset, c_str, size);

    ((Str_hd_t *)(str - sizeof (Str_hd_t)))->n_bytes = nb;
    return (str);
}

/***********************************************************************

    Returns the number of data bytes stored in "str".

***********************************************************************/

int STR_size (void *str) {
    if (str == NULL)
	return (0);
    return (((Str_hd_t *)((char *)str - sizeof (Str_hd_t)))->n_bytes);
}

/***********************************************************************

    Frees "str".

***********************************************************************/

void STR_free (void *str) {
    if (str != NULL)
        MISC_free ((char *)str - sizeof (Str_hd_t));
}

/***********************************************************************

    Expands the buffer of "str" to at least "length" bytes. Additional
    space is allocated in case the buffer must increase.

***********************************************************************/

static char *Str_expand (void *str, int length) {
    if (str != NULL &&
		((Str_hd_t *)(str - sizeof (Str_hd_t)))->b_size >= length)
	return (str);
    return (STR_grow (str, (length * 3) / 2));
}

/***************************************************************************

    Generates a string from a list of input strings. The arg list must be
    terminated by NULL. The first argument must be STR type which is used
    to place the generated string.

***************************************************************************/

char *STR_gen (char *str, ...) {
    va_list args;
    char *p;
    int cnt;

    va_start (args, str);
    cnt = 0;
    while (1) {
	p = va_arg (args, char *);
	if (p == NULL)
	    break;
	if (cnt == 0)
	    str = STR_copy (str, p);
	else
	    str = STR_cat (str, p);
	cnt++;
    }
    va_end (args);
    return (str);
}
